Fix for issue with initial-input in org-ql-view--complete-buffers-files (Fix #227)

ahmed-shariff opened the pull request.

alphapapa commented:
Which values have you tested this with?

ahmed-shariff commented:
For me the issues was when using the prefix argument for refreshing while in the org-ql-view/agenda buffer. I've tested this with different org file and queries. Is that what you were looking for? I am not familiar with writing tests on elisp.

ahmed-shariff commented:
An alternative is to use the `DEF` option of `completing-read`:
```elisp
(defun org-ql-view--complete-buffers-files ()
  "Return value for `org-ql-view-buffers-files' using completion."
  (cl-labels ((initial-input
               () (when org-ql-view-buffers-files
                    (org-ql-view--contract-buffers-files
                     org-ql-view-buffers-files))))
    (if (and org-ql-view-buffers-files
             (bufferp org-ql-view-buffers-files))
        ;; Buffers can't be input by name, so if the default value is a buffer, just use it.
        ;; TODO: Find a way to fix this.
        org-ql-view-buffers-files
      (org-ql-view--expand-buffers-files
       (completing-read "Buffers/Files: "
                        (list 'buffer 'org-agenda-files 'org-directory 'all)
                        ;; nil nil (initial-input))))))
                        nil nil nil nil (initial-input)))))) ;; (initial-input) is passed to DEF
```

alphapapa commented:
What I mean is, this is a matter of converting values to and from readable representations--it's a two-way conversion.  The variable in question may have a few different forms as its value, and each possible form must be printed and read again from the user's input.  You seem to be reporting a problem with one of those forms and trying to fix it for that form.  What about the other forms?

Using a different argument to `completing-read` is a different solution that brings its own issues.  I would prefer that users not have to take the extra step of `M-n` to get the value again.

ahmed-shariff commented:
I agree with your second point, that might not be ideal.

As for the different conditions, it seems I've misunderstood what how the function in question works. :)

Am I correct in assuming the `org-ql-view--complete-buffers-files`s behaviour is as follows:
-  if the contracted form of  `org-ql-view-buffers-files` is a string or `org-ql-view-buffers-files`  is nil -> prompt for input using `completing-read`
   - This would include the cases where the `org-ql-view-buffers-files` value corresponds to "buffer", "org-agenda-files",  "org-directory",  "org-agenda-files" and when it is `nil`
- if the `org-ql-view-buffers-files` is a buffer, simply return the buffer.
- if the `org-ql-view-buffers-files` is a list: this is the part that is causing the problem. I am not clear on how this value is expected to be handled? do you want to check if it is only one value, if so set it up with `completing-read`? or would it be better to simply return `org-ql-view-buffers-files` if the contracted form is not a string or `org-ql-view-buffers-files` is not nil?

Also is there a particular reason to wrap the `initial-input` form in a label?

I've pushed a version of `org-ql-view--complete-buffers-files` promts only  contracted form of  `org-ql-view-buffers-files` is a string or `org-ql-view-buffers-files`  is nil. Does this make more sense?

ahmed-shariff commented:
Also while digging into how the above function works, I came across this - the documentation of `org-ql-search` states that:
> - A space-separated list of file or buffer names

when using interactively. Is this supported? Testing it doesn't seem to work? I tried the following for "Buffer/files:"
- "\*scratch\* \*scratch2\*"
- "orgfile1.org orgfile2.org"

alphapapa commented:
> As for the different conditions, it seems I've misunderstood what how the function in question works. :)

The function exists because, e.g. if the user is searching the list of files in the variable `org-agenda-files`, that list is expanded  to the list of files when the search is run.  Afterward, if the user wants to modify the search, it would not be very helpful to have a long list of long filenames in the interactive completion, so, e.g. if the list of files is the same as that in `org-agenda-files`, the list of files is replaced with that variable name (and then converted back to the list of files when the search is run again).

I wasn't aware of any issue with `completing-read`, and it seems to work for me, so maybe this was caused by a change in Emacs.  What version are you using?

> Also is there a particular reason to wrap the initial-input form in a label?

`cl-labels` defines functions in a local scope.  Sometimes I use it to organize code more cleanly, like here.

> I've pushed a version of org-ql-view--complete-buffers-files promts only contracted form of org-ql-view-buffers-files is a string or org-ql-view-buffers-files is nil. Does this make more sense?

In order to fix this bug without potentially causing any others, I think it's necessary add tests for all of its intended input and output forms.  If we don't enumerate and test them, we're likely to overlook some.

> when using interactively. Is this supported? Testing it doesn't seem to work? I tried the following for "Buffer/files:"

What happens?

Anyway, try enclosing them in quotes.  If that still doesn't work, then it's probably a bug; I never pass buffer names interactively, so it's not something I've tested.  (This is an example of why it needs comprehensive tests.)

ahmed-shariff commented:
> I wasn't aware of any issue with completing-read, and it seems to work for me, so maybe this was caused by a change in Emacs. What version are you using?
I have 4 different version installed across different pc's I use :')
All >= 26

The issue happens with calling `org-ql-search` programmatically. For example if I eval the following:
```elisp
(org-ql-search '("~/Documents/orgfile1.org" "~/Documents/orgfile1.org")  query)
```
The query would run and I can get the view. 
But when I try to modify the query using `C-u r`, that's when the error I reported on the issue happens.

I just started looking into writing test cases in emacs. When I have some time, I'll try to write a few cases to test the expand and contract behaviour for the different conditions as well as for the `org-ql-view--complete-buffers-files` function.

To sum up, is the behavior of `org-ql-view--complete-buffers-files` defined as follows correct?
> `org-ql-view--complete-buffers-files` promts only if contracted form of `org-ql-view-buffers-files` is a string or `org-ql-view-buffers-files` is nil. 

I also opened a new issue for the space-seperated buffers/files issue #230.

alphapapa commented:
> But when I try to modify the query using `C-u r`, that's when the error I reported on the issue happens.

Does it also happen if you use the Transient dispatcher to modify the buffers/files list?  i.e. press `v i RET`.  I suspect it will.

> I just started looking into writing test cases in emacs. When I have some time, I'll try to write a few cases to test the expand and contract behaviour for the different conditions as well as for the org-ql-view--complete-buffers-files function.

Ok, thanks.  Please see `test-org-ql.el`, especially the parts that test the return values of certain functions (rather than the parts that test query results).

> To sum up, is the behavior of org-ql-view--complete-buffers-files defined as follows correct?      org-ql-view--complete-buffers-files promts only if contracted form of org-ql-view-buffers-files is a string or org-ql-view-buffers-files is nil.

No, I don't think that's correct.  Please see the function's definition; it's very short.

I think it's important for us to keep in mind that there are multiple functions interacting here: `org-ql-view--contract-buffers-files`, `org-ql-view--complete-buffers-files`, and `org-ql-view--expand-buffers-files`, each with its own distinct purpose.  That's why I think that tests are needed for each function before we can consider this issue (or issues) completely fixed.

ahmed-shariff commented:
> Does it also happen if you use the Transient dispatcher to modify the buffers/files list? i.e. press v i RET. I suspect it will.

Yep, same behavior.

I had pushed a set of test for the `org-ql-view--*-buffers-files` functions. I tried to account for their different use cases. But for `org-ql-view--complete-buffers-files` it assumes the definition I had shared before. Do you have any plans or ideas on how you want to handle the completion for entries that `org-ql-view--contract-buffers-files` doesn't contract: which are lists of non standard files names or buffer names, or buffers themselves? Would it be linked to implementing the issue in #230 ? The current behaviour on master is to return the value of `org-ql-view-buffers-files` if it is a buffer object without prompting for completion. I extended the same behaviour for non-standard lists. I am assuming that was a temporary solution?

alphapapa commented:
I'm impressed!  You learn and work quickly.  You seem to have made good use of Buttercup's features.  I haven't read every line of your tests yet, but it looks like you were thorough.

Note that buffers are not `read`able in Elisp, so the only way they could be completed or `read` is by name.  That leaves some potential ambiguity, as a string could be a filename or a buffer name.  A simple heuristic could be to check whether a string matches a buffer name (i.e. just call `get-buffer`): if it does, use the buffer, otherwise try `file-exists-p`; and if neither seems valid, signal with `user-error`.

What do you think?

Thanks very much for your work on this.  You've saved me a lot of time.  :)

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+              (spy-on 'org-agenda-files :and-return-value temp-filenames)
+              (expect (org-ql-view--contract-buffers-files temp-filenames) :to-equal "org-agenda-files"))
+            (it "org-directory"
+              (spy-on 'org-ql-search-directories-files :and-return-value temp-filenames)
```
I think it might be better to bind `org-directory` to the directory containing those files, rather than mocking `org-ql-search-directories-files`.  That way, that function is also tested indirectly, which could be helpful.

ahmed-shariff replied:
using let instead of spy

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+              (expect (org-ql-view--contract-buffers-files temp-filenames) :to-equal "org-agenda-files"))
+            (it "org-directory"
+              (spy-on 'org-ql-search-directories-files :and-return-value temp-filenames)
+              (spy-on 'org-agenda-files :and-return-value '())
```
Nitpick: since this is tested first in the function in question, it would be more, shall we say, symmetrical, to put it first in this test.  :)

ahmed-shariff replied:
Re-ordering :D

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+            (it "org-directory"
+              (spy-on 'org-ql-search-directories-files :and-return-value temp-filenames)
+              (spy-on 'org-agenda-files :and-return-value '())
+              (expect (org-ql-view--contract-buffers-files temp-filenames) :to-equal temp-filenames))
```
I may be wrong, because I wrote the code in question a long time ago, but I think this is supposed to return `"org-directory"` rather than the list of files.

ahmed-shariff replied:
you are right, this should be "org-directory"

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+                (expect (org-ql-view--contract-buffers-files value2) :to-equal value2))))
+          (describe "Expanding org-ql-view-buffers-files"
+            (it "all"
+              (let* ((buffer-names (list (make-temp-file "test" nil ".org") (make-temp-file "test" nil ".other")))
```
I'm a bit confused here: You're calling `make-temp-file`, which makes files on-disk, but then only using the filenames to create non-file-backed buffers.  Would `generate-new-buffer` or `generate-new-buffer-name` be more appropriate?

ahmed-shariff replied:
using `generate-new-buffer`. you learn something new everyday :D

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+            (it "all"
+              (let* ((buffer-names (list (make-temp-file "test" nil ".org") (make-temp-file "test" nil ".other")))
+                     (buffers (mapcar #'get-buffer-create buffer-names)))
+                (save-excursion
```
You could use `with-current-buffer` instead of `save-excursion/switch-to-buffer`.

ahmed-shariff replied:
using `with-current-buffer`

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+                (expect (org-ql-view--contract-buffers-files value1) :to-equal value1)
+                (expect (org-ql-view--contract-buffers-files value2) :to-equal value2))))
+          (describe "Expanding org-ql-view-buffers-files"
+            (it "all"
```
For my own edification, it would be helpful to add a clarifying comment like `;; all Org buffers`.  :)

ahmed-shariff replied:
Adding to descriptions

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+                (spy-on 'buffer-list :and-return-value buffers)
+                (expect (org-ql-view--expand-buffers-files "all") :to-equal (list (car buffers)))))
+            (it "org-agenda-files"
+              (spy-on 'org-agenda-files :and-return-value "value for org-agenda-files")
```
Mocking the function's value is probably fine, but it might be good to use a list value rather than a string, since that's what it's expected to return.

ahmed-shariff replied:
using let instead of spy

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+              (spy-on 'org-agenda-files :and-return-value "value for org-agenda-files")
+              (expect (org-ql-view--expand-buffers-files "org-agenda-files") :to-equal "value for org-agenda-files"))
+            (it "org-directory"
+              (spy-on 'org-ql-search-directories-files :and-return-value "value for org-directory")
```
Mocking the function's value is probably fine, but it might be good to use a list value rather than a string, since that's what it's expected to return.

ahmed-shariff replied:
using let instead of spy

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+              (spy-on 'org-ql-search-directories-files :and-return-value "value for org-directory")
+              (expect (org-ql-view--expand-buffers-files "org-directory") :to-equal "value for org-directory"))
+            (it "buffer"
+              (with-temp-buffer
```
Nice use of `with-temp-buffer`.  :)

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+            (it "literal values"
+              (with-temp-buffer
+                (expect (org-ql-view--expand-buffers-files (current-buffer)) :to-equal (current-buffer)))
+              (let ((test-buffer (get-buffer-create (make-temp-file "test"))))
```
As in the earlier comment, probably shouldn't call `make-temp-file` here.

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+                (expect (org-ql-view--expand-buffers-files test-list) :to-equal test-list))
+              (let ((string-literal "this is a string"))
+                (expect (org-ql-view--expand-buffers-files string-literal) :to-equal string-literal))))
+          (describe "Testing org-ql-view--complete-buffers-files"
```
Sometimes it seems awkward, but it would generally be good to write these test labels in the idiomatic Buttercup way, so that they read as sentence clauses in the output.  I'm not sure I've been entirely consistent with this, myself, but when done well, it makes test output much easier to understand.

ahmed-shariff replied:
Adding this to the descriptions

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+                (expect (org-ql-view--complete-buffers-files) :to-equal org-ql-view-buffers-files)
+                (expect 'completing-read :not :to-have-been-called)))
+            (it "buffer"
+              (let ((org-ql-view-buffers-files (get-buffer-create (make-temp-file "test"))))
```
As before, `make-temp-file`...  :)

alphapapa reviewed (changes requested):
Thanks again.

alphapapa commented:
You may be in the middle of working on this now, but I just happened to see this, and maybe it would be good to let you know now: I understand that factoring out that code into a function helps with testing, but I'd prefer not to do that unless it's really necessary.  I don't think we need to do that to test the behavior of these other functions.

alphapapa commented:
By the way, I've had second thoughts about mocking functions' return values.  I think we should generally avoid that, because it could mask problems if those functions change (and sometimes Org functions do).  Instead, we can bind the variables that affect those functions' return values.  (I'm making this comment here rather than as a review comment in each place in the code; sorry for any confusion.)

ahmed-shariff commented:
Anytime man. I enjoy hacking stuff, so :D
As I mentioned before I am new to writing tests as a whole, let alone on emacs. I'll make the changes you've requested and push it when I get some time. 

Regarding reading buffers, what you suggest makes sense. Following your suggestion, I have added the following:
- When the string passed to `org-ql-select` is not a file name but is a buffer name, return that. Previously, it would throw a user-error saying "can't open the file". I also moved the processing of the `buffers-of-files` in `org-ql-select` to a function outside. Just easier to test and also maybe can be used elsewhere if needed.
- When contracting `buffers-or-files` from the view, if it is a buffer, return the buffer name instead of the buffer object. This kind of solve the `todo` you had in `org-ql-view--complete-buffers-files`.

This still leaves the issue of how to process lists in the view. One potential solution is to define a separator (eg: ";") and allow typing in buffer-names/file-name separated by it. Is there a better way to do that?

ahmed-shariff commented:
Just saw the messages. I have a habit of abstracting whenever possible :)
I understand why you wouldn't want that as a separate function, I can move that back in. And also will make the fixes to the test as you mentioned.

ahmed-shariff commented:
Also, in the last commit, I am dumping the default value of the contracted `org-ql-view-buffers-files` as the initial-input (even list) and compare the output of `completing-read` to this value. If it is the same as the initial-input, return the previous value of `org-ql-view-buffers-files`. But this is a very hacky solution for when `org-ql-view-buffers-files` is a list. Coz typing something in with a similar syntax won't work.

alphapapa commented on org-ql.el:
```diff
@@ -367,6 +367,7 @@ returns nil or non-nil."
                                             ;; It feels unintuitive that `find-file-noselect' returns
                                             ;; a buffer if the filename doesn't exist.
                                             (find-file-noselect it))
+                                          (get-buffer it)
```
`org-ql-select` is not meant to be called interactively, and the docstring says that this value should be a list of buffers (not buffer names--admittedly, to Emacs, they are sometimes handled the same way).  So to support this, it should probably be done in `org-ql-search`'s `interactive` form.

ahmed-shariff replied:
The current implementation expects each `it` in that typcase to be a file. So if I pass something like `("*scrathc*" "*scratch2*") it results in the `user-error: Can't open file: %s` on the next line. Which is why I added this. This is not directly related to the issue at hand though.

ahmed-shariff replied:
The current implementation expects each `it` in that typcase to be a file. So if I pass something like `("*scrathc*" "*scratch2*")` it results in the `user-error: Can't open file: %s` on the next line. Which is why I added this. This is not directly related to the issue at hand though.

alphapapa commented:
> Coz typing something in with a similar syntax won't work.

Please give specific examples so we can talk about this concretely and try to devise a solution.  We might need to use `read` in some cases, but that raises its own issues (i.e. it would require users to quote every string, so we'd have to do that ourselves first, and then things start to get complicated).  Whatever we do, our goal here should be to be more explicit about what is supported, in what formats, and disallow inputs not in a valid format.

ahmed-shariff commented on org-ql-view.el:
```diff
@@ -1048,23 +1048,29 @@ current buffer.  Otherwise BUFFERS-FILES is returned unchanged."
        (buffer-file-name buffers-files))
+      ((pred bufferp)
+       (buffer-name buffers-files))
       (_ buffers-files))))
```
@alphapapa should this be made to throw an error?

ahmed-shariff commented:
> Maybe we should separate this work into two phases

I agree, but I am having a hard time separating these two:

To be clear, as I described before, the problem is when we have `org-ql-view-buffers-files` having a representation that the `org-ql-view--contract-buffers-files` doesn't provide a contracted form for, in the current case, these are lists that cannot be defined by `agenda-files` or `org-directory`. For instance if `org-ql-view-buffers-files` is set to `("orgfile1.org" "orgfile2.org)` what should the completion look like (for both the initial-input as well as when entering new values)?

I feel like the having the list represented as a space (or any other character (maybe make this a customizable value?)) separated string might be a better idea? What do you think?

On the same note, similar to my above comment, perhaps raise `user-error` if the input cannot be contracted to an expected value?

alphapapa commented:
> I feel like the having the list represented as a space (or any other character (maybe make this a customizable value?)) separated string might be a better idea? What do you think?

Filenames may contain spaces, so we have two choices: 1) quote every filename, which would be more tedious for the user; 2) only quote filenames with spaces, which would probably require either some kind of fancy parsing or calling `read` and converting unquoted filenames from symbols to strings.  Neither is ideal, but having to quote every filename, even ones without spaces, would be a poor UI, I think.

> 
> On the same note, similar to my above comment, perhaps raise `user-error` if the input cannot be contracted to an expected value/representation?

I think that would be confusing for the user: a query could work when calling `org-ql-search` but then signal an error when the user tries to edit the buffers-files variable.  I can imagine the bug reports...

Anyway, I think I'll plan for this issue/PR to be addressed/merged in 0.7, that way it won't hold up 0.6, which has had this issue for a while already without many users noticing.

ahmed-shariff commented:
> Anyway, I think I'll plan for this issue/PR to be addressed/merged in 0.7, that way it won't hold up 0.6, which has had this issue for a while already without many users noticing.

That sounds good.

> I think that would be confusing for the user: a query could work when calling org-ql-search but then signal an error when the user tries to edit the buffers-files variable. I can imagine the bug reports...

Should these two be consolidated? 

Regarding the multiple entries completion, here are a few other ideas:
- Using completing-read-multiple
- Have another default option the in completion list (eg: `enter custom values`) that leads to another read function like `read` or `completing-read-multiple`

Previously I kind of had issues with the `completing-read-multiple`, but since I moved to using selectrum, it's kind of grown on me. And I think it'd be a simpler (and cleaner?) solution and also allows to combine the default options as well (eg: `buffer` and `org-agenda-files`)

alphapapa commented:
> > I think that would be confusing for the user: a query could work when calling org-ql-search but then signal an error when the user tries to edit the buffers-files variable. I can imagine the bug reports...
> 
> Should these two be consolidated?

I'm not sure what you mean.

> Regarding the multiple entries completion, here are a few other ideas:
> 
>     * Using completing-read-multiple
> 
>     * Have another default option the in completion list (eg: `enter custom values`) that leads to another read function like `read` or `completing-read-multiple`
> 
> 
> Previously I kind of had issues with the `completing-read-multiple`, but since I moved to using selectrum, it's kind of grown on me. And I think it'd be a simpler (and cleaner?) solution and also allows to combine the default options as well (eg: `buffer` and `org-agenda-files`)

I haven't used completing-read-multiple much, so I don't know how well I like it.  But I think it's worth testing, yes.

ahmed-shariff commented:
> I'm not sure what you mean.

I meant the validation of the input. In retrospect, I think it might not be needed at all, as the values are processed by `org-ql-select` anyways.

I'll push a version that uses `completing-read-multiple`

ahmed-shariff commented:
@alphapapa would you be able to try the latest commit on this PR and see how the UI feels?

It uses the `completing-read-multiple` interface. By default the separator is ",", but it can be modified if needed by modifying `crm-separator`.
You should be able to use the values previously specified (`buffer` `org-agenda-files` `org-directory` `all`) and also any other buffer or file names.
For example, with the default separator the following is a valid value for the buffers/files prompt:
```
org-agenda-files,~/other_org_dir/org-file1.org,*scratch*
```

alphapapa commented:
Thanks, I think this will work pretty well.  A little feedback: here's another way to write this function that I think is a bit clearer:

```el
(defun org-ql-view--expand-buffers-files (buffers-files)
  "Return BUFFERS-FILES expanded to a list of files or buffers.
The counterpart to `org-ql-view--contract-buffers-files'."
  (--> (-list buffers-files)
    (pcase-exhaustive it
      ("all" (--select (equal (buffer-local-value 'major-mode it) 'org-mode)
                       (buffer-list)))
      ("org-agenda-files" (org-agenda-files))
      ("org-directory" (org-ql-search-directories-files))
      ((or "" "buffer") (current-buffer))
      ((pred bufferp) (list it))
      ((pred listp) (list it))
      ;; A single filename.
      ((pred stringp) (list it)))
    -non-nil -uniq -flatten))
```

ahmed-shariff commented:
Finally got around to wrapping this up.
I've added the functions and the test suites.

alphapapa commented on org-ql-view.el:
```diff
@@ -1032,54 +1032,91 @@ current buffer.  Otherwise BUFFERS-FILES is returned unchanged."
-      ((and (pred bufferp) (guard (buffer-file-name buffers-files)))
-       (buffer-file-name buffers-files))
-      (_ buffers-files))))
+    (-->
```
Wrapping a large form like this in `-->` is hard to read.  Please do something like bind the value and then make the `-->` form as small as possible.

alphapapa commented on org-ql-view.el:
```diff
@@ -1032,54 +1032,91 @@ current buffer.  Otherwise BUFFERS-FILES is returned unchanged."
+          ((pred bufferp)
+           (buffer-name buffers-files))
+          (_ buffers-files))
+      ;; All items needs to be strings to pick duplicates when used with the extend conterpart.
+      ;; So making sure the buffers are convered to file names
```
This comment has a typo and could be a bit more concise.  :)

alphapapa commented on org-ql-view.el:
```diff
@@ -1032,54 +1032,91 @@ current buffer.  Otherwise BUFFERS-FILES is returned unchanged."
+          (_ buffers-files))
+      ;; All items needs to be strings to pick duplicates when used with the extend conterpart.
+      ;; So making sure the buffers are convered to file names
+      (if (stringp it)
+          it
+        (-map
+         (lambda (buffer-file)
+           (if (bufferp buffer-file)
+               (--if-let (buffer-file-name buffer-file)
+                   it
+                 (buffer-name buffer-file))
+             buffer-file))
+         it)))))
```
This form is very hard to follow with its deep nesting and multiple levels of `if`.  Please rewrite it to be flatter, e.g. using `cond`, `cl-typecase`, etc.

alphapapa commented on org-ql-view.el:
```diff
@@ -1032,54 +1032,91 @@ current buffer.  Otherwise BUFFERS-FILES is returned unchanged."
-    ((pred stringp) buffers-files)))
+The counterpart to `org-ql-view--contract-buffers-files'.
+This always returns a list of string values."
+  (-->
+      (-map (lambda (buffer-file)
+                (pcase-exhaustive buffer-file
+                  ("all" (--select (equal (buffer-local-value 'major-mode it) 'org-mode)
+                                   (buffer-list)))
+                  ("org-agenda-files" (org-agenda-files))
+                  ("org-directory" (org-ql-search-directories-files))
+                  ((or "" "buffer")
+                   (current-buffer))
+                  ((pred bufferp) (list buffer-file))
+                  ;; A single filename.
+                  ((pred stringp) (list buffer-file))
+                  (_ (error (format "Value %s is not a valid buffer/file" buffer-file)))))
+            (-list buffers-files))
+    -flatten -non-nil
+    ;; expanding all file-buffers to file names to avoid duplicate entries being formed
+    (-map (lambda (buffer-file)
+            (if (bufferp buffer-file)
+                (--if-let (buffer-file-name buffer-file)
+                    it
+                  (buffer-name buffer-file))
+              buffer-file))
+          it)
+    -uniq))
```
As with the other form, this is too large of an initial value for a `-->` form.

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -1686,19 +1686,19 @@ with keyword arg NOW in PLIST."
           (it "Errors for an unquoted lambda"
             (expect (open-link unquoted-lambda-link)
-                    :to-throw 'error '("CAUTION: Link not opened because unsafe buffers-files parameter detected: (lambda nil (error UNSAFE))")))
+                    :to-throw 'error '("Value lambda is not a valid buffer/file")))
           (it "Errors for a quoted lambda in a list"
             (if (version< (org-version) "9.3")
                 (expect (open-link quoted-lambda-in-list-link)
-                        :to-throw 'error '("CAUTION: Link not opened because unsafe buffers-files parameter detected: ((quote (lambda nil (error UNSAFE))))"))
+                        :to-throw 'error '("Value (quote (lambda nil (error UNSAFE))) is not a valid buffer/file"))
               (expect (open-link quoted-lambda-in-list-link)
-                      :to-throw 'error '("CAUTION: Link not opened because unsafe buffers-files parameter detected: ('(lambda nil (error UNSAFE)))"))))
+                      :to-throw 'error '("Value (quote (lambda nil (error UNSAFE))) is not a valid buffer/file"))))
           (it "Errors for an unquoted lambda in a list"
             (expect (open-link unquoted-lambda-in-list-link)
-                    :to-throw 'error '("CAUTION: Link not opened because unsafe buffers-files parameter detected: ((lambda nil (error UNSAFE)))"))))
+                    :to-throw 'error '("Value (lambda nil (error UNSAFE)) is not a valid buffer/file"))))
```
Why did you change these errors? These are important safety tests.

ahmed-shariff replied:
I had added a explicit error when the input to the `org-ql-view--expand-buffers-files` function, 
https://github.com/alphapapa/org-ql/blob/564d491604f76aa05e6e0a8e9d87b762153a78fd/org-ql-view.el#L1108

According to my understanding, the previous errors being tested were because of the `pcase` failing. So roll back the error I triggering and the `pacse` to fail which would lead to the older errors?
Even then, in this implementation the return value is always a list, which makes the error being produced to be a bit different than what it previously had.

alphapapa replied:
I'm not sure what you mean.  Do you understand the potential security issue here?

ahmed-shariff replied:
To clarify, the potential values for the `org-ql-view--expand-buffers-files` now are the following:
- A buffer
- A string
- A file-name
- one of "all", "org-agenda-files", "org-directory", "buffer", or an empty string (which evaluates to the same values as "buffer")
- a list containing above values

I am also converting all buffer objects to string values to avoid duplicates that are introduced because of the buffer/buffer-name of a file and a file-name being passed. 

As a result this function now always returns a list of strings (or at least should, will add tests for these :D).

This is in part due to moving to using `completing-read-multiple` which returns a list by default. Because of this, I had dropped the `listp` predicate in the `pcase` inside the function since now all elements in the list are being processed individually. This in-turn results in the unsafe values for buffers/files being tested in the link safety section never getting to the point where that error gets signalled since they never match any of the values in the current pcase. Which was why I was signalling an error when an invalid value is passed. Does this make sense?

ahmed-shariff replied:
I may have overlooked that functions are a valid input to `buffers-files` in `org-ql-select`. Which might be creating a whole bunch of new problems :)

Currently (in the master branch) the following should create a view with the function
```elisp
(org-ql-search #'a-function-returning-list-of-files '(level 1))
```
with the value for `org-ql-view-buffers-files` is set to `#'a-function-returning-list-of-files`. 

If I understand this correctly, this is left as such so refreshing the view would recall the function? (or alternatively the evaluated value can be stored?)

I am not sure how a function would translate to a value in `completing-read-mulitple`, so I am skipping prompting for that entirely when refreshing a view with a function as a value for `org-ql-view-buffers-files`.

The safety concern you were referring to before comes from allowing a function to be used as a parameter for `org-ql-select`? Currently the `org-ql-view--expand-buffers-files` function would signal an error if used with a function, and when storing a link, the function's evaled value is used for `buffers-files`.

Does that cover the problem with unsafe values for `buffers-files`?

alphapapa replied:
You can see where the error is signaled here: https://github.com/alphapapa/org-ql/blob/94f9e6f3031b32cf5e2149beca7074807235dcb0/org-ql-view.el#L638

That safety check should be left there, in that function, where links are followed.  IIUC, the change you made to the tests would cause that check to not be tested, which could lead to security vulnerabilities in the future.

ahmed-shariff replied:
That check looks at either the value returned by the `org-ql-view--expand-buffers-files` or the current-buffer. If I understand it correctly, it was there to make sure the value returned by the `org-ql-view--expand-buffers-files` function is not another function that can get evaluated once passed to `org-ql-select`. Currently, if I understand when and how `org-ql-view--expand-buffers-files` gets called, in a valid situation, it only should get values that I had previously mentioned, anything else would result in the error on the new tests being signaled.

If the concern is future changes to the `org-ql-view--expand-buffers-files` function results in returning a function value, that function can be mocked to make sure the error in https://github.com/alphapapa/org-ql/blob/94f9e6f3031b32cf5e2149beca7074807235dcb0/org-ql-view.el#L638
get's signaled?

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+              (expect (org-ql-view--contract-buffers-files 'org-agenda-files) :to-equal "org-agenda-files")
+              (expect (org-ql-view--contract-buffers-files #'org-agenda-files) :to-equal "org-agenda-files"))
+            (it "arbitarary list of buffers/files"
+              (let ((value1 '("a.org" "b.org"))
+                    (value2 'a))
```
Nitpick: `value1` and `value2` aren't very descriptive variable names.  Could we use something more meaningful, like `before`/`after`, or something like that?

alphapapa commented on tests/test-org-ql.el:
```diff
@@ -2111,7 +2111,84 @@ with keyword arg NOW in PLIST."
+              (let ((value1 '("a.org" "b.org"))
+                    (value2 'a))
+                (expect (org-ql-view--contract-buffers-files value1) :to-equal value1)
+                ;; If the value does not result to a buffer, file, or string, throws error
```
Nitpick: This isn't grammatical.  How about something like, "Signals error if value is not a buffer, file, or string."

(Note also the use of "signal" rather than "throw": in Lisp, `throw` is something else.)

Generally, please check the comments in this PR and ensure they're complete, grammatical sentences, ending in a period.  :)

alphapapa reviewed (changes requested):
Thanks.  This is a quick first-pass review.  After this review is addressed, I'll do a more thorough one when I have time.

alphapapa commented:
BTW, it would be easier for me to review the code and the changes made in response if you left the conversations "unresolved" until the changes are actually made.  When I come back to a large change like this after a while, it takes some time to reacquaint myself with it and what I said about it before.

ahmed-shariff commented:
I'm new to the review process in git :D

Regarding the error signals being different when testing for safety of the link parameters.

To clarify, the potential values for the `org-ql-view--expand-buffers-files` now are the following:
- A buffer
- A string
- A file-name
- one of "all", "org-agenda-files", "org-directory", "buffer", or an empty string (which evaluates to the same values as "buffer")
- a list containing above values

I am also converting all buffer objects to string values to avoid duplicates that are introduced because of the buffer/buffer-name of a file and a file-name being passed. 

As a result this function now always returns a list of strings (or at least should, will add tests for these :D).

This is in part due to moving to using `completing-read-multiple` which returns a list by default. Because of this, I had dropped the `listp` predicate in the `pcase` inside the function since now all elements in the list are being processed individually. This in-turn results in the unsafe values for buffers/files being tested in the link safety section never getting to the point where that error gets signalled since they never match any of the values in the current pcase. Which was why I was signalling an error when an invalid value is passed. Does this make sense?

alphapapa commented on org-ql-view.el:
```diff
@@ -1093,30 +1108,37 @@ representation `org-ql-view-buffers-files' is returned."
-              buffer-file))
-          it)
-    -uniq))
+  (let ((expanded-buffers-files
+         (--> buffers-files
+           -list -non-nil
+           (-map (lambda (buffer-file)
+                   (pcase-exhaustive buffer-file
+                     ("all" (--select (equal (buffer-local-value 'major-mode it) 'org-mode)
+                                      (buffer-list)))
+                     ("org-agenda-files" (org-agenda-files))
+                     ("org-directory" (org-ql-search-directories-files))
+                     ((or "" "buffer")
+                      (current-buffer))
+                     ((or (pred bufferp)
+                          ;; A single filename.
+                          (pred stringp))
+                      buffer-file)
+                     (_ (error (format "Value %s is not a valid buffer/file" buffer-file)))))
+                 it))))
+    (--> expanded-buffers-files
+      -flatten
+      ;; removing `nil' again as some values have been expanded.
+      -non-nil
+      ;; Expanding all buffers to file names or buffer names to remove duplicate entries.
+      (--map
+       (pcase-exhaustive it
+         ((pred bufferp) (or (buffer-file-name it)
+                             (buffer-name it)))
+         ;; Any values at this point should be a buffer or string.
+         ;; Testing for string anyways.
+         ((pred stringp) it))
+       it)
+      -uniq)))
```
This is still a very long, narrow form that is very hard to follow.  Maybe you could rewrite it in a functional way using `cl-labels`.

alphapapa commented on org-ql-view.el:
```diff
@@ -1032,39 +1035,48 @@ current buffer.  Otherwise BUFFERS-FILES is returned unchanged."
+                                       (buffer-name buffer-file))))
+                 it)
+                -uniq))
+        (t (error (format "Value %s is not a string, a valid function or a list of buffer/strings" contracted-buffers-files)))))))
```
This is still a very long form that's hard to follow.  Please try to rewrite it in a more functional way.

ahmed-shariff commented:
I totally understand. Ill close this PR and start anew.

ahmed-shariff commented:
if it's any consolation, I also need a few minutes every time I revisit this XD
